openlr/openlr_match_quality/openlr_assessment_tool/traffic_mode.cpp

```diff
@@ -1,144 +1,210 @@
+#include <QMessageBox>
 
-#include <QtCore/QItemSelection>
+#include <tuple>
 
+namespace
+{
+std::vector<m2::PointD> GetPoints(openlr::LinearSegment const & segment)
+{
+  std::vector<m2::PointD> result;
+  for (auto const & lrp : segment.m_locationReference.m_points)
+    result.push_back(MercatorBounds::FromLatLon(lrp.m_latLon));
+  return result;
+}
+
+// TODO(mgsergio): Optimize this.
+std::vector<m2::PointD> GetReachablePoints(m2::PointD const & srcPoint,
+                                           std::vector<m2::PointD> const path,
+                                           PointsControllerDelegateBase const & pointsDelegate,
+                                           size_t const lookBackIndex)
+{
+  auto reachablePoints = pointsDelegate.GetReachablePoints(srcPoint);
+  if (lookBackIndex < path.size())
+  {
+    auto const & toBeRemoved = path[path.size() - lookBackIndex - 1];
+    reachablePoints.erase(
+        remove_if(begin(reachablePoints), end(reachablePoints),
+                  [&toBeRemoved](m2::PointD const & p) {
+                    return p.EqualDxDy(toBeRemoved, 1e-6);
+                  }),
+        end(reachablePoints));
+  }
+  return reachablePoints;
+}
+}
 
-// DecodedSample -----------------------------------------------------------------------------------
-DecodedSample::DecodedSample(Index const & index, openlr::SamplePool const & sample)
+namespace impl
 {
-  for (auto const & item : sample)
+// static
+size_t const RoadPointCandidate::kInvalidId = std::numeric_limits<size_t>::max();;
+
+/// This class denotes a "non-deterministic" feature point.
+/// I.e. it is a set of all pairs <FeatureID, point index>
+/// located at a specified coordinate.
+/// Only one point at a time is considered active.
+RoadPointCandidate::RoadPointCandidate(std::vector<FeaturePoint> const & points,
+                                       m2::PointD const & coord)
+  : m_coord(coord)
+  , m_candidates(points)
+{
+  LOG(LDEBUG, ("Candidate points:", points));
+}
+
+void RoadPointCandidate::ActivatePoint(RoadPointCandidate const & rpc)
+{
+  for (auto const & fp1 : m_candidates)
   {
-    m_decodedItems.push_back(item);
-    for (auto const & mwmSegment : item.m_segments)
+    for (auto const & fp2 : rpc.m_candidates)
     {
-      auto const & fid = mwmSegment.m_fid;
-      Index::FeaturesLoaderGuard g(index, fid.m_mwmId);
-      CHECK(fid.m_mwmId.IsAlive(), ("Mwm id is not alive."));
-      if (m_points.find(fid) != end(m_points))
-        continue;
-
-      FeatureType ft;
-      CHECK(g.GetFeatureByIndex(fid.m_index, ft), ("Can't read feature", fid));
-      ft.ParseGeometry(FeatureType::BEST_GEOMETRY);
-
-      auto & v = m_points[fid];
-      v.reserve(ft.GetPointsCount());
-      ft.ForEachPoint(MakeBackInsertFunctor(v), scales::GetUpperScale());
+      if (fp1.first == fp2.first)
+      {
+        SetActivePoint(fp1.first);
+        return;
+      }
     }
   }
+  CHECK(false, ("One mutual featrue id should exist."));
```

Comment:
I believe there should be exactly one.
```
             c
             •----------f2----
            /
 a •--f1---• b
            \
             •----------f3----
             d
```

We start from a and got b. At point b there are 3 candidate features, at a there is one. So when going from a to b at be there must be exactly one mutual feature (f1) with point a.